File: backend/controllers/course_controller.py
# controllers/course_controller.py
from typing import List, Dict, Any, Tuple
from .base_controller import BaseController
from models.course import Course
from models.department import Department
from services.course_service import CourseService

class CourseController(BaseController):
    """Controller for course-related operations"""
    
    def __init__(self):
        super().__init__()
        self.service = CourseService()
    
    def get_courses_by_department(self, department_id: int) -> Tuple[Dict, int]:
        """Get all courses for a department"""
        try:
            courses = self.service.get_courses_by_department(department_id)
            return self.success_response(data=courses)
        except Exception as e:
            return self.handle_exception(e, "Failed to retrieve courses")
    
    def get_course(self, course_id: int) -> Tuple[Dict, int]:
        """Get a specific course with details"""
        try:
            course = self.service.get_course_with_details(course_id)
            if not course:
                return self.error_response("Course not found", 404)
            
            return self.success_response(data=course)
        except Exception as e:
            return self.handle_exception(e, "Failed to retrieve course")
    
    def get_course_equivalents(self, course_id: int) -> Tuple[Dict, int]:
        """Get equivalent courses for a given course"""
        try:
            equivalents = self.service.get_course_equivalents(course_id)
            return self.success_response(data=equivalents)
        except Exception as e:
            return self.handle_exception(e, "Failed to retrieve course equivalents")
    
    def search_courses(self) -> Tuple[Dict, int]:
        """Search courses based on criteria"""
        try:
            query = self.request.args.get('q', '')
            institution_id = self.request.args.get('institution_id', type=int)
            department_id = self.request.args.get('department_id', type=int)
            
            courses = self.service.search_courses(
                query=query,
                institution_id=institution_id,
                department_id=department_id
            )
            
            return self.success_response(data=courses)
        except Exception as e:
            return self.handle_exception(e, "Failed to search courses")
    
    def create_course(self) -> Tuple[Dict, int]:
        """Create a new course"""
        try:
            data = self.request.get_json()
            if not data:
                return self.error_response("No data provided")
            
            result = self.service.create_course(data)
            
            if 'errors' in result:
                return self.validation_error_response(result['errors'])
            
            return self.success_response(
                data=result['course'],
                message="Course created successfully",
                status_code=201
            )
        except Exception as e:
            return self.handle_exception(e, "Failed to create course")

File: frontend/src/controllers/CourseController.js
// controllers/CourseController.js
import { Course } from '../models/Course.js';
import { ApiService } from '../services/api.js';

export class CourseController {
    constructor() {
        this.apiService = new ApiService();
    }

    async getCoursesByDepartment(departmentId) {
        try {
            const response = await this.apiService.get(`/courses?department_id=${departmentId}`);
            return {
                success: true,
                data: response.data.map(course => Course.fromAPI(course))
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    async getCourseEquivalents(courseId) {
        try {
            const response = await this.apiService.get(`/equivalents?course_id=${courseId}`);
            return {
                success: true,
                data: response.data.map(course => Course.fromAPI(course))
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    async searchCourses(query, institutionId = null, departmentId = null) {
        try {
            const params = new URLSearchParams();
            if (query) params.append('q', query);
            if (institutionId) params.append('institution_id', institutionId);
            if (departmentId) params.append('department_id', departmentId);

            const response = await this.apiService.get(`/courses/search?${params}`);
            return {
                success: true,
                data: response.data.map(course => Course.fromAPI(course))
            };
        } catch (error) {
            return {
                success: false,
                error: error.message
            };
        }
    }

    toggleCourseSelection(course, selectedCourses, setSelectedCourses) {
        const isSelected = selectedCourses.some(c => c.id === course.id);
        if (isSelected) {
            setSelectedCourses(selectedCourses.filter(c => c.id !== course.id));
        } else {
            setSelectedCourses([...selectedCourses, course]);
        }
    }
}